File: packages/python/elizaos/bootstrap/providers/relationships.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("RELATIONSHIPS")


def format_relationship(
    relationship: dict[str, str | int | list[str] | dict[str, str]],
    target_name: str,
) -> str:
    tags = relationship.get("tags", [])
    tags_str = (", ".join(tags) if tags else "none") if isinstance(tags, list) else str(tags)

    interactions = relationship.get("metadata", {})
    interaction_count = interactions.get("interactions", 0) if isinstance(interactions, dict) else 0

    return f"- {target_name}: tags=[{tags_str}], interactions={interaction_count}"


async def get_relationships(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    entity_id = message.entity_id
    if not entity_id:
        return ProviderResult(
            text="No relationships found.",
            values={"relationshipCount": 0},
            data={"relationships": []},
        )

    try:
        relationships = await runtime.get_relationships(entity_id=entity_id)
    except Exception as e:
        runtime.logger.debug(
            {"src": "provider:relationships", "agentId": runtime.agent_id, "error": str(e)},
            "Failed to get relationships",
        )
        relationships = []

    if not relationships:
        return ProviderResult(
            text="No relationships found.",
            values={"relationshipCount": 0},
            data={"relationships": []},
        )

    sorted_relationships = sorted(
        relationships,
        key=lambda r: (r.get("metadata", {}) or {}).get("interactions", 0),
        reverse=True,
    )[:30]

    formatted_relationships: list[str] = []
    entity_cache: dict[object, str] = {}
    for rel in sorted_relationships:
        target_id = rel.get("targetEntityId")
        if not target_id:
            continue

        target_name = entity_cache.get(target_id)
        if target_name is None:
            target_entity = await runtime.get_entity(target_id)
            target_name = target_entity.name if target_entity else str(target_id)[:8]
            entity_cache[target_id] = target_name

        formatted_relationships.append(format_relationship(rel, target_name))

    if not formatted_relationships:
        return ProviderResult(
            text="No relationships found.",
            values={"relationshipCount": 0},
            data={"relationships": []},
        )

    sender_name = message.content.sender_name if message.content else "Unknown"
    text = f"# {runtime.character.name} has observed {sender_name} interacting with:\n" + "\n".join(
        formatted_relationships
    )

    return ProviderResult(
        text=text,
        values={
            "relationshipCount": len(sorted_relationships),
        },
        data={
            "relationships": sorted_relationships,
        },
    )


relationships_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_relationships,
    dynamic=_spec.get("dynamic", True),
)

File: packages/plugin-bootstrap/src/providers/relationships.ts
import type { Entity, IAgentRuntime, Memory, Provider, Relationship, UUID } from '@elizaos/core';

/**
 * Escape a value for CSV output.
 * Wraps in quotes if contains comma, quote, or newline.
 */
function csvEscape(value: string | number | null | undefined): string {
  if (value === null || value === undefined) return '';
  const str = String(value);
  if (str.includes(',') || str.includes('"') || str.includes('\n')) {
    return `"${str.replace(/"/g, '""')}"`;
  }
  return str;
}

/**
 * Formats relationships as CSV for token efficiency.
 * Format: name,interactions,tags
 *
 * @param runtime - The runtime instance
 * @param relationships - The relationships to format
 * @returns CSV formatted string
 */
async function formatRelationships(runtime: IAgentRuntime, relationships: Relationship[]) {
  // Sort relationships by interaction strength (descending)
  const sortedRelationships = relationships
    .filter((rel) => rel.metadata?.interactions)
    .sort(
      (a, b) =>
        ((b.metadata?.interactions as number | undefined) || 0) -
        ((a.metadata?.interactions as number | undefined) || 0)
    )
    .slice(0, 30); // Get top 30

  if (sortedRelationships.length === 0) {
    return '';
  }

  // Deduplicate target entity IDs to avoid redundant fetches
  const uniqueEntityIds = Array.from(
    new Set(sortedRelationships.map((rel) => rel.targetEntityId as UUID))
  );

  // Fetch all required entities in a single batch operation
  const entities = await Promise.all(uniqueEntityIds.map((id) => runtime.getEntityById(id)));

  // Create a lookup map for efficient access
  const entityMap = new Map<string, Entity | null>();
  entities.forEach((entity, index) => {
    if (entity) {
      entityMap.set(uniqueEntityIds[index], entity);
    }
  });

  // CSV header
  const rows: string[] = ['name,interactions,tags'];

  // Format relationships as CSV rows
  for (const rel of sortedRelationships) {
    const targetEntityId = rel.targetEntityId as UUID;
    const entity = entityMap.get(targetEntityId);

    if (!entity) continue;

    const name = entity.names[0] || 'Unknown';
    const interactions = (rel.metadata?.interactions as number) || 0;
    const tags = rel.tags?.join(';') || '';

    rows.push(`${csvEscape(name)},${interactions},${csvEscape(tags)}`);
  }

  return rows.join('\n');
}

/**
 * Provider for fetching relationships data.
 *
 * @type {Provider}
 * @property {string} name - The name of the provider ("RELATIONSHIPS").
 * @property {string} description - Description of the provider.
 * @property {Function} get - Asynchronous function to fetch relationships data.
 * @param {IAgentRuntime} runtime - The agent runtime object.
 * @param {Memory} message - The message object containing entity ID.
 * @returns {Promise<Object>} Object containing relationships data or error message.
 */
const relationshipsProvider: Provider = {
  name: 'RELATIONSHIPS',
  description:
    'Relationships between {{agentName}} and other people, or between other people that {{agentName}} has observed interacting with',
  dynamic: true,
  get: async (runtime: IAgentRuntime, message: Memory) => {
    // Get all relationships for the current user
    const relationships = await runtime.getRelationships({
      entityId: message.entityId,
    });

    if (!relationships || relationships.length === 0) {
      return {
        data: {
          relationships: [],
        },
        values: {
          relationships: 'No relationships found.',
        },
        text: 'No relationships found.',
      };
    }

    const formattedRelationships = await formatRelationships(runtime, relationships);

    if (!formattedRelationships) {
      return {
        data: {
          relationships: [],
        },
        values: {
          relationships: 'No relationships found.',
        },
        text: 'No relationships found.',
      };
    }

    const senderName = message.content.senderName || message.content.name || 'user';
    const header = `# Relationships (${relationships.length}) - ${senderName}'s connections`;

    return {
      data: {
        relationships: formattedRelationships,
      },
      values: {
        relationships: formattedRelationships,
      },
      text: `${header}\n${formattedRelationships}`,
    };
  },
};

export { relationshipsProvider };
